pyquil/api/_config.py

```diff
@@ -93,65 +135,164 @@ class PyquilConfig(object):
         "default": None
     }
 
-    def __init__(self):
-        self.configparsers = {}
-        for env_name, default_path in CONFIG_PATHS.items():
+    def __init__(self, config_paths: Dict[str, str] = CONFIG_PATHS):
+        """
+        :param config_paths: the paths to the various configuration files read by PyQuil
+        """
+
+        # The engagement callback can be added by config consumers after construction
+        self.get_engagement = lambda: None
+
+        self.config_parsers = {}
+        for env_name, default_path in config_paths.items():
             default_path = expanduser(default_path)
             path = environ.get(env_name, default_path)
 
             cp = ConfigParser()
             cp.read(abspath(path))
-            self.configparsers[env_name] = cp
+            self.config_parsers[env_name] = cp
+        self._parse_auth_tokens()
 
-    def _env_or_config_or_default(self, env=None, file=None, section=None, name=None, default=None):
+    def _parse_auth_tokens(self):
+        self.user_auth_token = _parse_auth_token(self.user_auth_token_path, ['access_token', 'refresh_token', 'scope'])
+        self.qmi_auth_token = _parse_auth_token(self.qmi_auth_token_path, ['access_token', 'refresh_token'])
+
+    def _env_or_config_or_default(self,
+                                  env=None,
+                                  file=None,
+                                  section=None,
+                                  name=None,
+                                  engagement_key=None,
+                                  default=None):
         """
         Get the value of the environment variable or config file value.
         The environment variable takes precedence.
 
         :param env: The environment variable name.
         :param name: The config file key.
+        :param engagement_key: The attribute name by which this value can be read from an engagement.
+            If None, then this value is not provided by engagement.
         :return: The value or None if not found
         """
+
+        # If it's an envvar, that overrides all other options
         env_val = environ.get(env)
         if env_val is not None:
             return env_val
 
+        # Otherwise, use the values in the config files from disk
         try:
-            return self.configparsers[file].get(section, name)
+            return self.config_parsers[file].get(section, name)
         except (NoSectionError, NoOptionError, KeyError):
-            return default
+            pass
+
+        # If no local configuration is available, certain values are provided
+        #   by the dispatch service.
+        try:
+            if engagement_key is not None and self.engagement is not None:
+                return getattr(self.engagement, engagement_key)
+        except AttributeError:
+            pass
+
+        return default
 
     @property
-    def api_key(self):
-        return self._env_or_config_or_default(**self.API_KEY)
```

Comment:
should we be deprecating this?

Reply:
Do you think so? The service won't accept it any more and we don't have any need to read it in